tools/qontract_cli.py

```diff
@@ -3944,85 +3932,176 @@ def remove(ctx, sso_client_vault_secret_path: str):
-    vault_settings = get_app_interface_vault_settings()
-    secret_reader = create_secret_reader(use_vault=vault_settings.vault)
-    er_settings = get_settings()[0]
-    m_inventory = load_module_inventory(get_modules())
-    factories = setup_factories(er_settings, m_inventory, er_inventory, secret_reader)
-    f = factories.get_factory(spec.provision_provider)
-    resource = f.create_external_resource(spec)
-    f.validate_external_resource(resource)
-    print(resource.json(exclude={"data": {FLAG_RESOURCE_MANAGED_BY_ERV2}}))
+    print(erv2cli.input_data)
 
 
 @external_resources.command()
-@click.argument("provision-provider", required=True)
-@click.argument("provisioner", required=True)
-@click.argument("provider", required=True)
-@click.argument("identifier", required=True)
 @click.pass_context
-def request_reconciliation(
-    ctx, provision_provider: str, provisioner: str, provider: str, identifier: str
-):
+def request_reconciliation(ctx):
     """Marks a resource as it needs to get reconciled. The itegration will reconcile the resource at
-    its next iteration.
+    its next iteration."""
+    erv2cli = Erv2Cli(
+        provision_provider=ctx.obj["provision_provider"],
+        provisioner=ctx.obj["provisioner"],
+        provider=ctx.obj["provider"],
+        identifier=ctx.obj["identifier"],
+        secret_reader=ctx.obj["secret_reader"],
+    )
+    erv2cli.reconcile()
+
+
+@external_resources.command()
+@binary(["terraform"])
+@binary_version("terraform", ["version"], TERRAFORM_VERSION_REGEX, TERRAFORM_VERSION)
+@click.option(
+    "--dry-run/--no-dry-run",
+    help="Enable/Disable dry-run. Default: dry-run enabled!",
+    default=True,
+)
+@click.pass_context
+def migrate(ctx, dry_run: bool) -> None:
+    """Migrate an existing external resource managed by terraform-resources to ERv2.
+
 
-    e.g: e.g: qontract-reconcile --config=<config> external-resources request-reconciliation aws app-sre-stage rds dashdotdb-stage
+    E.g: qontract-reconcile --config=<config> external-resources migrate aws app-sre-stage rds dashdotdb-stage
     """
-    er_settings = get_settings()[0]
-    vault_settings = get_app_interface_vault_settings()
-    secret_reader = create_secret_reader(use_vault=vault_settings.vault)
-    with get_aws_api(
-        query_func=gql.get_api().query,
-        account_name=er_settings.state_dynamodb_account.name,
-        region=er_settings.state_dynamodb_region,
-        secret_reader=secret_reader,
-    ) as aws_api:
-        state_manager = ExternalResourcesStateDynamoDB(
-            aws_api=aws_api,
-            table_name=er_settings.state_dynamodb_table,
-        )
-        key = ExternalResourceKey(
-            provision_provider=provision_provider,
-            provisioner_name=provisioner,
-            provider=provider,
-            identifier=identifier,
+    if ctx.obj["provider"] == "rds":
+        # The "random_password" is not an AWS resource. It's just in the outputs and can't be migrated :(
```

Comment:
RandomPassword has its own resource using the RandomProvider (not just in the outputs). I understand it could be migrated (At least it's worth a try)

Reply:
I tried the migration without the `random_password` resource, but it doesn't work. In the `terraform-resources` state, there isn't any reference to a `random_password` resource. All I can see is `resource.instances[].aws_db_instance.password` has the previously generated password set.

```
Generated Terraform code for the stacks: CDKTF

Initializing the backend...

Successfully configured the backend "s3"! Terraform will automatically
use this backend unless the backend configuration changes.

Initializing provider plugins...
- Finding hashicorp/aws versions matching "5.60.0"...
- Finding hashicorp/random versions matching "3.6.2"...
- Using hashicorp/aws v5.60.0 from the shared cache directory
- Using hashicorp/random v3.6.2 from the shared cache directory

Terraform has created a lock file .terraform.lock.hcl to record the provider
selections it made above. Include this file in your version control repository
so that Terraform can guarantee to make the same selections by default when
you run "terraform init" in the future.


Warning: Incomplete lock file information for providers

Due to your customized provider installation methods, Terraform was forced to
calculate lock file checksums locally for the following providers:
  - hashicorp/aws
  - hashicorp/random

The current .terraform.lock.hcl file only includes checksums for linux_amd64,
so Terraform running on another platform will fail to install these
providers.

To calculate additional checksums for another platform, run:
  terraform providers lock -platform=linux_amd64
(where linux_amd64 is the platform to generate)

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
CDKTF  Initializing the backend...
CDKTF  Initializing provider plugins...
       - Reusing previous version of hashicorp/aws from the dependency lock file
CDKTF  - Reusing previous version of hashicorp/random from the dependency lock file
CDKTF  - Using previously-installed hashicorp/aws v5.60.0
CDKTF  - Using previously-installed hashicorp/random v3.6.2

       Terraform has been successfully initialized!
CDKTF  You may now begin working with Terraform. Try running "terraform plan" to see
       any changes that are required for your infrastructure. All Terraform commands
       should now work.

       If you ever set or change modules or backend configuration for Terraform,
       rerun this command to reinitialize your working directory. If you forget, other
       commands will detect it and remind you to do so if necessary.
CDKTF  aws_db_instance.rds-migration-01 (rds-migration-01): Refreshing state... [id=db-D5OSETCHOKZJLDRBVYDHWOIRCQ]
CDKTF  Note: Objects have changed outside of Terraform

       Terraform detected the following changes made outside of Terraform since the
       last "terraform apply" which may have affected this plan:

         # aws_db_instance.rds-migration-01 (rds-migration-01) has changed
         ~ resource "aws_db_instance" "rds-migration-01" {
             + db_name                               = "postgres"
               id                                    = "db-D5OSETCHOKZJLDRBVYDHWOIRCQ"
               tags                                  = {
                   "app"                    = "msk-example"
                   "cluster"                = "appint-ex-01"
                   "environment"            = "production"
                   "managed_by_integration" = "terraform_resources"
                   "namespace"              = "example-msk-01"
               }
               # (45 unchanged attributes hidden)
           }


       Unless you have made equivalent changes to your configuration, or ignored the
       relevant attributes using ignore_changes, the following plan may include
       actions to undo or respond to these changes.

       ───────────────────────────────────────────────────────────────────────────────

       Terraform used the selected providers to generate the following execution plan.
       Resource actions are indicated with the following symbols:
         + create
         ~ update in-place

       Terraform will perform the following actions:
CDKTF  # aws_db_instance.rds-migration-01 (rds-migration-01) will be updated in-place
         ~ resource "aws_db_instance" "rds-migration-01" {
             + allow_major_version_upgrade           = false
               id                                    = "db-D5OSETCHOKZJLDRBVYDHWOIRCQ"
             ~ password                              = (sensitive value)
             ~ tags                                  = {
                   "app"                    = "msk-example"
                   "cluster"                = "appint-ex-01"
                   "environment"            = "production"
                 ~ "managed_by_integration" = "terraform_resources" -> "external_resources"
                   "namespace"              = "example-msk-01"
               }
             ~ tags_all                              = {
                 ~ "managed_by_integration" = "terraform_resources" -> "external_resources"
                   # (4 unchanged elements hidden)
               }
               # (53 unchanged attributes hidden)
           }

         # random_password.rds-migration-01-password (rds-migration-01-password) will be created
         + resource "random_password" "rds-migration-01-password" {
             + bcrypt_hash = (sensitive value)
             + id          = (known after apply)
             + keepers     = {
                 + "reset_password" = ""
               }
             + length      = 20
             + lower       = true
             + min_lower   = 0
             + min_numeric = 0
             + min_special = 0
             + min_upper   = 0
             + number      = true
             + numeric     = true
             + result      = (sensitive value)
             + special     = true
             + upper       = true
           }

       Plan: 1 to add, 1 to change, 0 to destroy.

       Changes to Outputs:
         + rds-migration-01-rds__db_host     = "rds-migration-01.cllmpvoboklu.us-east-1.rds.amazonaws.com"
         + rds-migration-01-rds__db_name     = "postgres"
         + rds-migration-01-rds__db_password = (sensitive value)
         + rds-migration-01-rds__db_port     = 5432
         + rds-migration-01-rds__db_user     = (sensitive value)
CDKTF  random_password.rds-migration-01-password (rds-migration-01-password): Creating...
CDKTF  random_password.rds-migration-01-password (rds-migration-01-password): Creation complete after 0s [id=none]
CDKTF  aws_db_instance.rds-migration-01 (rds-migration-01): Modifying... [id=db-D5OSETCHOKZJLDRBVYDHWOIRCQ]
CDKTF  Error: updating RDS DB Instance (rds-migration-01): operation error RDS: ModifyDBInstance, https response error StatusCode: 400, RequestID: d83ffefc-5b33-4663-a032-8f44c2b1c265, api error InvalidParameterValue: The parameter MasterUserPassword is not a valid password. Only printable ASCII characters besides '/', '@', '"', ' ' may be used.

         with aws_db_instance.rds-migration-01 (rds-migration-01),
         on cdk.tf.json line 90, in resource.aws_db_instance.rds-migration-01 (rds-migration-01):
         90:       }


0 Stacks deploying     1 Stack done     0 Stacks waiting
Invoking Terraform CLI failed with exit code 1
```